balancer/rls/internal/cache/cache.go

```diff
@@ -0,0 +1,237 @@
+	// amount of time to backoff, given the number of retries.
+	Backoff backoff.Strategy
+	// Timer fires when the backoff period ends and incoming requests after
+	// this will trigger a new RLS request.
+	Timer *time.Timer
+	// Callback provided by the LB policy to be notified when the backoff timer
+	// expires. This will trigger a new picker to be returned to the
+	// ClientConn, to force queued up RPCs to be retried.
+	Callback func()
+}
+
+// lruEntry is the actual entry which is stored in the LRU cache. It requires
+// the key (in addition to the actual entry) as well, since the onEvicted
+// callback and expiry timer would need the key to perform their job.
+type lruEntry struct {
+	key  Key
+	val  *Entry
+	size int
+}
+
+// LRU is a cache with a least recently used eviction policy.
+//
+// It is not safe for concurrent access. The RLS LB policy will provide a mutex
+// which will synchronize access to this cache from its multiple users: the LB
+// policy, picker and the expiry timer.
+type LRU struct {
+	maxSize   int
+	usedSize  int
+	onEvicted func(Key, *Entry)
+
+	ll    *list.List
+	cache map[Key]*list.Element
+}
+
+// NewLRU creates a cache.LRU with a size limit of maxSize and the provided
+// eviction callback.
+//
+// Currently, only the cache.Key and the HeaderData field from cache.Entry
+// count towards the size of the cache (other overhead per cache entry is not
+// counted). The cache could temporarily exceed the configured maxSize because
+// we want the entries to spend a configured minimum amount of time in the
+// cache before they are LRU evicted (so that all the work performed in sending
+// an RLS request and caching the response is not a total waste).
+//
+// The provided onEvited callback must not attempt to re-add the entry inline
+// and the RLS LB policy does not have a need to do that.
+//
+// The cache package trusts the RLS policy (its only user) to supply a default
+// minimum non-zero maxSize, in the event that the ServiceConfig does not
+// provide a value for it.
+func NewLRU(maxSize int, onEvicted func(Key, *Entry)) *LRU {
+	return &LRU{
+		maxSize:   maxSize,
+		onEvicted: onEvicted,
+		ll:        list.New(),
+		cache:     make(map[Key]*list.Element),
+	}
+}
+
+// TODO(easwars): If required, make this function more sophisticated.
+func entrySize(key Key, value *Entry) int {
+	return len(key.Path) + len(key.KeyMap) + len(value.HeaderData)
+}
+
+// removeToFit removes older entries from the cache to make room for a new
+// entry of size newSize.
+func (lru *LRU) removeToFit(newSize int) {
+	for lru.usedSize+newSize > lru.maxSize {
+		elem := lru.ll.Back()
+		if elem == nil {
+			// This is a corner case where the cache is empty, but the new entry
+			// to be added is bigger than maxSize.
+			grpclog.Info("rls: newly added cache entry exceeds cache maxSize")
+			return
+		}
+
+		entry := elem.Value.(*lruEntry).val
+		if t := entry.EarliestEvictTime; !t.IsZero() && t.Before(time.Now()) {
```

Comment:
time.Now() may be slow-ish; let's sample it once before the `for` loop?

Reply:
Done.